adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,496 @@
+not yet supported.
+
+[[table.cuda.CUDA_features_to_SYCL]]
+.CUDA execution features with their corresponding SYCL features
+[width="100%",options="header",cols="50%,50%"]
+|====
+| [code]#SYCL#                                                       | [code]#CUDA#
+| [code]#Compute unit#                                               | [code]#Streaming multiprocessor#
+| [code]#nd_range#                                                   | [code]#grid#
+| [code]#work-group#                                                 | [code]#Thread block#
+| [code]#sub-group#                                                  | [code]#thread_block_tile#
+| [code]#work-item#                                                  | [code]#Thread#
+| [code]#SYCL nd_item synchronization#                               | [code]#cudaStreamSynchronize#
+| [code]#work-group barrier#                                         | [code]#__syncthread#
+|====
+
+[[sec::programming_interface]]
+== Programming Interface
+
+[[sub:cuda:application_interoperability]]
+=== Application Interoperability
+
+This section describes the API level interoperability between SYCL and CUDA.
+
+The CUDA backend supports API interoperability for `platform`, `device`,
+`context`, `queue`, `event` and `buffer`. Interoperability for `kernel`,
+`kernel_bundle`, `device_image`, `sampled_image` and `unsampled_image` is not
+supported.
+
+[[table.cuda.appinterop.nativeobjects]]
+.Types of native backend objects application interoperability
+[width="100%",options="header",cols="20%,20%,20%,40%"]
+|====
+| [code]#SyclType# | [code]#backend_input_t<backend::cuda, SyclType># | [code]#backend_return_t<backend::cuda, SyclType># | Description
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts , however it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue encapsulates a CUDA stream.
+| [code]#event#    | `CUevent`    | `CUevent`  | A SYCL event encapsulates a CUDA event.
+| [code]#buffer# | `struct { CUdeviceptr ptr; size_t size; }` | `CUdeviceptr` | A SYCL buffer encapsulates a CUDA device pointer.
+|====
+
+[[table.cuda.appinterop.make_interop_APIs]]
+.[code]#make_*# Interoperability APIs for native backend objects.
+[width="100%",options="header",cols="40%,60%"]
+|====
+| CUDA interoperability function                                    |  Description
+| [code]#template<backend Backend> +
+platform + 
+make_platform(const backend_input_t<Backend, platform> &backendObject);# 
+        | Create a SYCL `platform` from a list of CUDA device, the list must contain at least one CUDA device.
```

Comment:
Only one platform makes sense but I'm not sure it's a requirement we want to introduce. For example at the moment DPC++ will return one platform per CUDA device, it will likely be changed in the future to only one platform, but still it's possible for vendors to organize their platforms in different ways.

It does make it a little strange if someone tries to create a platform with a subset of devices that's not in `get_platforms()`. So we might need to revisit this a bit.

Reply:
I think it is already a requirement because the SYCL core spec says the following in [section 4.6.2 "Platform class"](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:platform-class):

> The execution environment for a SYCL application has a fixed number of platforms which does not vary as the application executes. The application can get a list of all these platforms via platform::get_platforms(). The platform class also provides constructors, but constructing a new platform instance merely creates a new object that is a copy of one of the objects returned by platform::get_platforms().

Reply:
Well yes, the fixed number of platforms is a requirement, what I meant was that I'm not sure the backend spec should mandate that all CUDA devices be under the same platform.

Reply:
Yes, I think that makes sense too.  I'd just like to see some wording in the description of `make_platform()` that the returned `platform` object is a copy of one of the objects returned by the core API `platform::get_platforms()`.  For example, this is what the Level Zero interop spec says:

> Constructs a SYCL platform instance from a Level-Zero `ze_driver_handle_t`. The SYCL execution environment contains a fixed number of platforms that are enumerated via `sycl::platform::get_platforms()`. Calling this function does not create a new platform. Rather it merely creates a `sycl::platform` object that is a copy of one of the platforms from that enumeration.

Reply:
The wording for this has been updated to reflect that `make_platform` doesn't create a new platform just returns a `platform` object that is a copy of one of the existing platforms, so I think this is resolved now.